File: scripts/terraform.py
#!/usr/bin/env python3

from glob import glob
import shlex
import tempfile
import os
import sys
import time
import itertools
import os.path
import subprocess


class Terraform(object):
    def __init__(self):
        self.var_dir = 'vars'
        self.shared_env = 'common'
        self._var_files = None
        self.verbosity = 0
        self.skip_prompt = False

    def main(self):
        import argparse
        argparser = argparse.ArgumentParser()
        argparser.add_argument('env')
        argparser.add_argument('--verbose', '-v', dest='verbosity', action='count', default=0)
        argparser.add_argument('--skip-prompt', action='store_true',
                               help='Skip workspace switch prompt')
        subparsers = argparser.add_subparsers()

        plan_parser = subparsers.add_parser('plan')
        plan_parser.add_argument('--destroy', action='store_true')
        plan_parser.add_argument('--out')
        plan_parser.add_argument('--target', action='append')
        plan_parser.add_argument('--apply', dest='apply_plan', action='store_true')
        plan_parser.add_argument('--refresh-only', action='store_true')
        plan_parser.set_defaults(parser_func=self.plan)

        apply_parser = subparsers.add_parser('apply')
        apply_parser.add_argument("--plan")
        apply_parser.add_argument("--target")
        apply_parser.add_argument("--approve", action='store_true')
        apply_parser.add_argument('--refresh-only', action='store_true')
        apply_parser.set_defaults(parser_func=self.apply)

        destroy_parser = subparsers.add_parser('destroy')
        destroy_parser.add_argument("--approve", action='store_true')
        destroy_parser.set_defaults(parser_func=self.destroy)

        show_parser = subparsers.add_parser('show')
        show_parser.add_argument("--plan")
        show_parser.set_defaults(parser_func=self.show)

        graph_parser = subparsers.add_parser('graph')
        graph_parser.add_argument("--draw-cycles", action='store_true')
        graph_parser.set_defaults(parser_func=self.graph)

        init_parser = subparsers.add_parser('init')
        init_parser.add_argument("--upgrade", action='store_true')
        init_parser.add_argument("--backend-config",
                                 help='A backend config file or key=value assignment',
                                 nargs='+')
        init_parser.set_defaults(parser_func=self.init)

        new_workspace_parser = subparsers.add_parser('new-workspace')
        new_workspace_parser.set_defaults(parser_func=self.new_workspace)

        taint_parser = subparsers.add_parser('taint')
        taint_parser.add_argument("--module", help='Module containing the resource to taint')
        taint_parser.add_argument("name", help="A resource to taint", nargs='+')
        taint_parser.set_defaults(parser_func=self.taint)

        untaint_parser = subparsers.add_parser('untaint')
        untaint_parser.add_argument("--module", help='Module containing the resource to untaint')
        untaint_parser.add_argument("name", help="A resource to untaint", nargs='+')
        untaint_parser.set_defaults(parser_func=self.untaint)

        import_parser = subparsers.add_parser('import')
        import_parser.add_argument("--allow-missing-config", action='store_true',
                                   help='Allow import when no resource configuration block exists.')
        import_parser.add_argument("addr", help="Address to import resource to")
        import_parser.add_argument("resource_id", help="Resource-specific ID")
        import_parser.set_defaults(parser_func=self.import_action)

        terraform_parser = subparsers.add_parser('terraform')
        terraform_parser.add_argument("args", help="A resource to untaint", nargs='+')
        terraform_parser.set_defaults(parser_func=self.terraform)

        migration_fromplan_parser = subparsers.add_parser('migration_fromplan')
        migration_fromplan_parser.add_argument("tfplan_file", help="input from terraform plan -out")
        migration_fromplan_parser.add_argument("migration_file", help="output HCL migration file")
        migration_fromplan_parser.set_defaults(parser_func=self.migration_fromplan)

        migrate_plan_parser = subparsers.add_parser('migrate_plan')
        migrate_plan_parser.add_argument("migration_file", help="HCL migration file")
        migrate_plan_parser.set_defaults(parser_func=self.migrate_plan)

        migrate_apply_parser = subparsers.add_parser('migrate_apply')
        migrate_apply_parser.add_argument("migration_file", help="HCL migration file")
        migrate_apply_parser.set_defaults(parser_func=self.migrate_apply)

        args = argparser.parse_args()

        self.env = args.env
        self.verbosity = args.verbosity
        self.skip_prompt = args.skip_prompt

        if self._verbosity_is(2):
            print("args: {}".format(args))

        if 'parser_func' not in vars(args):
            # no environment given
            sys.stderr.write('no sub-command given\n\n')
            argparser.print_help()
            exit(1)

        args_ignore = set(['parser_func', 'env', 'verbosity', 'skip_prompt'])
        kwargs = {k: v for k, v in vars(args).items() if k not in args_ignore}
        args.parser_func(**kwargs)

    def init(self, backend_config=None, upgrade=None):
        self._tfenv_init()
        args = []
        if backend_config is not None:
            for config in backend_config:
                args.append('-backend-config')
                args.append(config)
        if upgrade:
            args.append("--upgrade")
        rc = self._script(
            # no switching workspaces for init -- not necessary
            ['terraform', 'init'] + list(args))
        exit(rc)

    def new_workspace(self):
        self._tfenv_init()
        rc = self._script(['terraform', 'workspace', 'new', self.env])
        exit(rc)

    def terraform(self, args):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        cmd = ['terraform']
        cmd.extend(args)
        rc = self._script(
            self._tf_workspace_select(),
            cmd)
        if rc != 0:
            exit(rc)

    def plan(self, destroy=False, out=None, apply_plan=False, target=None, refresh_only=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        plan_path = out
        if apply_plan and plan_path is None:
            plan_dir = ".tf-plans"
            os.makedirs(plan_dir, exist_ok=True)
            plan_name_prefix = "tf-plan-{}-{}-".format(self.env, int(time.time()))
            plan_file, plan_path = tempfile.mkstemp(
                prefix=plan_name_prefix,
                suffix=".out",
                dir=plan_dir)
            os.close(plan_file)

        base_args = ['terraform', 'plan']
        var_file_args = self._var_file_args()
        extra_args = []
        if destroy:
            extra_args.append('-destroy')
        if plan_path is not None:
            arg = '-out={}'.format(plan_path)
            extra_args.append(arg)
        if target is not None:
            for t in target:
                extra_args.extend(['-target', t])
        if refresh_only:
            extra_args.append('-refresh-only')
        args = itertools.chain(base_args, var_file_args, extra_args)
        rc = self._script(
            self._tf_workspace_select(),
            args)
        if apply_plan:
            if rc != 0:
                print("Aborted -- planning failed")
                exit(rc)
            print()
            print("Would you like to apply this plan?")
            cont = ''
            while cont != 'yes':
                print("You must answer 'yes' to continue")
                try:
                    cont = input('> ')
                except EOFError:
                    print()
                    cont = 'no'
                if cont == 'no':
                    print("Aborted -- plan will not be applied")
                    print("path: {}".format(plan_path))
                    exit(1)
            print()
            self.apply(plan=plan_path)
        exit(rc)

    def apply(self, plan=None, target=None, refresh_only=None, approve=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        args = []
        if plan:
            args = [plan]
        else:
            args = list(self._var_file_args())
        if target is not None:
            args.extend(['-target', target])
        if refresh_only:
            args.append('-refresh-only')
        if approve:
            args.append('-auto-approve')
        rc = self._script(
            self._tf_workspace_select(),
            ['terraform', 'apply'] + list(args))
        exit(rc)

    def destroy(self, approve=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        args = list(self._var_file_args())
        if approve:
            args.append('-auto-approve')
        rc = self._script(
            self._tf_workspace_select(),
            ['terraform', 'destroy'] + list(args))
        exit(rc)

    def show(self, plan=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        args = []
        if plan:
            args = [plan]
        rc = self._script(
            self._tf_workspace_select(),
            ['terraform', 'show'] + list(args))
        exit(rc)

    def graph(self, draw_cycles=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        args = []
        if draw_cycles:
            args = ['-draw-cycles']
        rc = self._script(
            self._tf_workspace_select(),
            ['terraform', 'graph'] + list(args))
        exit(rc)

    def import_action(self, allow_missing_config=None, addr=None, resource_id=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        base_args = ['terraform', 'import']
        var_file_args = self._var_file_args()
        extra_args = []
        if allow_missing_config:
            extra_args.extend(['-allow-missing-config'])
        args = itertools.chain(base_args, var_file_args, extra_args, [addr, resource_id])
        rc = self._script(
            self._tf_workspace_select(),
            args)
        exit(rc)

    def taint(self, name=None, module=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        names = name
        if isinstance(name, str):
            names = [name]
        cmd_base = ['terraform', 'taint']
        if module is not None:
            cmd_base.extend(['-module', module])
        for n in names:
            cmd = cmd_base + [n]
            rc = self._script(
                self._tf_workspace_select(),
                cmd)
            if rc != 0:
                exit(rc)

    def untaint(self, name=None, module=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        names = name
        if isinstance(name, str):
            names = [name]
        cmd_base = ['terraform', 'untaint']
        if module is not None:
            cmd_base.extend(['-module', module])
        for n in names:
            cmd = cmd_base + [n]
            rc = self._script(
                self._tf_workspace_select(),
                cmd)
            if rc != 0:
                exit(rc)

    def migration_fromplan(self, tfplan_file=None, migration_file=None):
        self._tfenv_init()
        show_cmd = ("terraform", "show", "-json", tfplan_file)
        tfedit_cmd = ("tfedit", "migration", "fromplan", "-o="+migration_file)
        tfshow = subprocess.Popen(show_cmd, stdout=subprocess.PIPE)
        subprocess.check_call(tfedit_cmd, stdin=tfshow.stdout)
        tfshow.wait()

    def migrate_plan(self, migration_file=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        var_file_args = " ".join(self._var_file_args())
        cmd = ("tfmigrate", "plan", migration_file)
        environ = os.environ.copy()
        environ["TFMIGRATE_LOG"] = "DEBUG"
        environ["TF_CLI_ARGS_plan"] = var_file_args
        environ["TF_CLI_ARGS_import"] = var_file_args
        subprocess.check_call(self._tf_workspace_select())
        subprocess.check_call(cmd, env=environ)

    def migrate_apply(self, migration_file=None):
        self._tfenv_init()
        self._check_env()
        self._prompt_env_switch()
        var_file_args = " ".join(self._var_file_args())
        cmd = ("tfmigrate", "apply", migration_file)
        environ = os.environ.copy()
        environ["TFMIGRATE_LOG"] = "DEBUG"
        environ["TF_CLI_ARGS_plan"] = var_file_args
        environ["TF_CLI_ARGS_apply"] = var_file_args
        environ["TF_CLI_ARGS_import"] = var_file_args
        subprocess.check_call(self._tf_workspace_select())
        subprocess.check_call(cmd, env=environ)

    def _check_env(self):
        '''
        We want to fail most commands if the environment doesn't exist.  In the past this was
        harmless but under new deployment methods you can end up in the wrong workspace for the
        subproject you are in.
        '''
        env_dir = os.path.join(self.var_dir, self.env)
        if not os.path.exists(env_dir):
            sys.stderr.write('\nNo environment in this project: {}\n\n.'.format(self.env))
            exit(1)

    def _prompt_env_switch(self):
        # FIXME:  Get the real current environment
        curr_env = self._exec_capture('terraform', 'workspace', 'show')
        curr_env = curr_env.decode('utf-8').strip()
        if curr_env == self.env:
            return
        sys.stderr.write(
                '\nswitching environment \x1b[31m{}\x1b[0m ~> \x1b[32m{}\x1b[0m\n\n'.format(curr_env, self.env))
        if self.skip_prompt:
            return
        while 1:
            sys.stderr.write("switch to {}? [y/N] ".format(self.env))
            resp = input()
            if resp:
                resp = resp.lower()
            else:
                resp = 'n'
            if resp == 'n' or resp == 'no':
                exit(1)
            if resp == 'y' or resp == 'yes':
                break
            sys.stderr.write('what?\n')

    def _tfenv_init(self):
        if not os.path.exists('.terraform-version'):
            # There is no specified version so we will use the default installed in the container.
            warn = 'WARNING: .terraform-version not found -- using the default installed terraform'
            sys.stderr.write('\n' + warn + '\n\n')
            return
        rc = self._script(["tfenv", "install"])
        if rc != 0:
            exit(rc)

    def _locate_var_files(self):
        if self._var_files is None:
            common_files = self._all_var_files(os.path.join(self.var_dir, self.shared_env))
            common_files = [os.path.join(self.shared_env, path) for path in common_files]
            env_files = self._all_var_files(os.path.join(self.var_dir, self.env))
            env_files = [os.path.join(self.env, path) for path in env_files]
            self._var_files = common_files + env_files
            if self._verbosity_is(1):
                self._log_var_files()
        return (os.path.join(self.var_dir, rel_path) for rel_path in self._var_files)

    def _all_var_files(self, d):
            pattern = os.path.join(d, '*.tfvars')
            return [os.path.basename(path) for path in glob(pattern)]

    def _log_var_files(self):
        for name in self._var_files:
            print("Variable file: {}".format(name))

    def _var_file_args(self):
        paths = list(self._locate_var_files())
        return ("-var-file={}".format(path) for path in paths)

    def _verbosity_is(self, level):
        return self.verbosity >= level

    def _tf_workspace_select(self, name=None):
        if name is None:
            name = self.env
        return ['terraform', 'workspace', 'select', name]

    def _script(self, *cmds, **kwargs):
        def mkcmd(args):
            return ' '.join(shlex.quote(a) for a in args)
        script = ' && '.join(map(mkcmd, cmds))
        chdir = kwargs.get('chdir')
        if chdir:
            cd = 'cd {}'.format(shlex.quote(chdir))
            script = '{} && {}'.format(cd, script)
        cmd = '({})'.format(script)
        sys.stderr.write(cmd + '\n')
        if kwargs.get('dry_run'):
            return 0
        return self._sh(cmd)

    def _exec(self, *cmdargs, **kwargs):
        cmd = ' '.join(shlex.quote(a) for a in cmdargs)
        chdir = kwargs.get('chdir')
        if chdir:
            cmd = '(cd {} && {})'.format(chdir, cmd)
        sys.stderr.write(cmd + '\n')
        return self._sh(cmd)

    def _script_capture(self, *cmds, **kwargs):
        def mkcmd(args):
            return ' '.join(shlex.quote(a) for a in args)
        script = ' && '.join(map(mkcmd, cmds))
        chdir = kwargs.get('chdir')
        cmd = '({})'.format(script)
        sys.stderr.write(cmd + '\n')
        return subprocess.check_output(['sh', '-c', script], shell=True, cwd=chdir)

    def _exec_capture(self, *cmdargs, **kwargs):
        cmd = ' '.join(shlex.quote(a) for a in cmdargs)
        chdir = kwargs.get('chdir')
        sys.stderr.write(cmd + '\n')
        return subprocess.check_output(cmd, shell=True, cwd=chdir)

    def _sh(self, cmd):
        process = subprocess.Popen(['/bin/bash', '-c', cmd], stdin=sys.stdin, stderr=sys.stderr)
        process.communicate()
        rc = process.wait()
        return rc

if __name__ == '__main__':
    tf = Terraform()
    tf.main()

File: internal/terraform/terraform.go
package terraform

import (
	"bufio"
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/luthersystems/mars/internal/app"
	"github.com/luthersystems/mars/internal/runner"
)

const tfenvInstallLockPath = "/opt/tfenv/versions/.install.lock"

type InitCmd struct {
	BackendConfig []string `name:"backend-config" help:"A backend config file or key=value assignment."`
	Upgrade       bool     `name:"upgrade" help:"Upgrade modules and plugins."`
	Reconfigure   bool     `name:"reconfigure" help:"Reconfigure the backend on init."`
}

type NewWorkspaceCmd struct{}

type RawCmd struct {
	Args []string `arg:"" optional:"" passthrough:"" name:"args"`
}

type PlanCmd struct {
	Destroy     bool     `name:"destroy"`
	Out         string   `name:"out"`
	Target      []string `name:"target"`
	ApplyPlan   bool     `name:"apply"`
	RefreshOnly bool     `name:"refresh-only" xor:"refresh"`
	SkipRefresh bool     `name:"skip-refresh" xor:"refresh"`
}

type ApplyCmd struct {
	Plan        string `name:"plan"`
	Target      string `name:"target"`
	Approve     bool   `name:"approve"`
	RefreshOnly bool   `name:"refresh-only"`
}

type DestroyCmd struct {
	Approve bool `name:"approve"`
}

type ShowCmd struct {
	Plan string `name:"plan"`
}

type GraphCmd struct {
	DrawCycles bool `name:"draw-cycles"`
}

type ImportCmd struct {
	AllowMissingConfig bool   `name:"allow-missing-config" help:"Allow import when no resource configuration block exists."`
	Addr               string `arg:"" help:"Address to import resource to."`
	ResourceID         string `arg:"" name:"resource_id" help:"Resource-specific ID."`
}

type TaintCmd struct {
	Module string   `name:"module" help:"Module containing the resource to taint."`
	Name   []string `arg:"" name:"name" help:"A resource to taint."`
}

type UntaintCmd struct {
	Module string   `name:"module" help:"Module containing the resource to untaint."`
	Name   []string `arg:"" name:"name" help:"A resource to untaint."`
}

type MigrationFromPlanCmd struct {
	TFPlanFile    string `arg:"" name:"tfplan_file" help:"Input from terraform plan -out."`
	MigrationFile string `arg:"" name:"migration_file" help:"Output HCL migration file."`
}

type MigratePlanCmd struct {
	MigrationFile string `arg:"" name:"migration_file" help:"HCL migration file."`
}

type MigrateApplyCmd struct {
	MigrationFile string `arg:"" name:"migration_file" help:"HCL migration file."`
}

type service struct {
	env        string
	verbosity  int
	skipPrompt bool
	runner     runner.Runner
	stdin      io.Reader
	stdout     io.Writer
	stderr     io.Writer
	clock      app.Clock
}

func (c *InitCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.tfenvInit(ctx); err != nil {
		return err
	}
	args := []string{"terraform", "init"}
	for _, config := range c.BackendConfig {
		args = append(args, "-backend-config", config)
	}
	if c.Upgrade {
		args = append(args, "--upgrade")
	}
	if c.Reconfigure {
		args = append(args, "--reconfigure")
	}
	return s.sequence(ctx, runner.Cmd(args[0], args[1:]...))
}

func (c *NewWorkspaceCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.tfenvInit(ctx); err != nil {
		return err
	}
	return s.sequence(ctx, runner.Cmd("terraform", "workspace", "new", s.env))
}

func (c *RawCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	if len(c.Args) > 0 && c.Args[0] == "--" {
		c.Args = c.Args[1:]
	}
	cmd := append([]string{"terraform"}, c.Args...)
	return s.sequence(ctx, s.workspaceSelect(), runner.Cmd(cmd[0], cmd[1:]...))
}

func (c *PlanCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	planPath := c.Out
	if c.ApplyPlan && planPath == "" {
		if err := os.MkdirAll(".tf-plans", 0o755); err != nil {
			return err
		}
		clock := s.clock
		if clock == nil {
			clock = realClock{}
		}
		file, err := os.CreateTemp(".tf-plans", fmt.Sprintf("tf-plan-%s-%d-*.out", s.env, clock.Unix()))
		if err != nil {
			return err
		}
		planPath = file.Name()
		if err := file.Close(); err != nil {
			return err
		}
	}

	args := append([]string{"terraform", "plan"}, s.varFileArgs()...)
	if c.Destroy {
		args = append(args, "-destroy")
	}
	if planPath != "" {
		args = append(args, "-out="+planPath)
	}
	for _, target := range c.Target {
		args = append(args, "-target", target)
	}
	if c.RefreshOnly {
		args = append(args, "-refresh-only")
	}
	if c.SkipRefresh {
		args = append(args, "-refresh=false")
	}

	if err := s.sequence(ctx, s.workspaceSelect(), runner.Cmd(args[0], args[1:]...)); err != nil {
		if c.ApplyPlan {
			fmt.Fprintln(s.stdout, "Aborted -- planning failed")
		}
		return err
	}
	if !c.ApplyPlan {
		return nil
	}
	if !s.confirmApply(planPath) {
		return runner.Exit(1)
	}
	return (&ApplyCmd{Plan: planPath}).Run(ctx, rt)
}

func (c *ApplyCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	var args []string
	if c.Plan != "" {
		args = []string{c.Plan}
	} else {
		args = s.varFileArgs()
	}
	if c.Target != "" {
		args = append(args, "-target", c.Target)
	}
	if c.RefreshOnly {
		args = append(args, "-refresh-only")
	}
	if c.Approve {
		args = append(args, "-auto-approve")
	}
	return s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", append([]string{"apply"}, args...)...))
}

func (c *DestroyCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	args := s.varFileArgs()
	if c.Approve {
		args = append(args, "-auto-approve")
	}
	return s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", append([]string{"destroy"}, args...)...))
}

func (c *ShowCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	args := []string{"show"}
	if c.Plan != "" {
		args = append(args, c.Plan)
	}
	return s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", args...))
}

func (c *GraphCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	args := []string{"graph"}
	if c.DrawCycles {
		args = append(args, "-draw-cycles")
	}
	return s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", args...))
}

func (c *ImportCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	args := append([]string{"import"}, s.varFileArgs()...)
	if c.AllowMissingConfig {
		args = append(args, "-allow-missing-config")
	}
	args = append(args, c.Addr, c.ResourceID)
	return s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", args...))
}

func (c *TaintCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	for _, name := range c.Name {
		args := []string{"taint"}
		if c.Module != "" {
			args = append(args, "-module", c.Module)
		}
		args = append(args, name)
		if err := s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", args...)); err != nil {
			return err
		}
	}
	return nil
}

func (c *UntaintCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	for _, name := range c.Name {
		args := []string{"untaint"}
		if c.Module != "" {
			args = append(args, "-module", c.Module)
		}
		args = append(args, name)
		if err := s.sequence(ctx, s.workspaceSelect(), runner.Cmd("terraform", args...)); err != nil {
			return err
		}
	}
	return nil
}

func (c *MigrationFromPlanCmd) Run(ctx context.Context, rt *app.Runtime) error {
	s := newService(rt)
	if err := s.tfenvInit(ctx); err != nil {
		return err
	}
	return s.runner.Pipe(
		ctx,
		runner.Cmd("terraform", "show", "-json", c.TFPlanFile),
		runner.Cmd("tfedit", "migration", "fromplan", "-o="+c.MigrationFile),
	)
}

func (c *MigratePlanCmd) Run(ctx context.Context, rt *app.Runtime) error {
	return migrate(ctx, rt, "plan", c.MigrationFile, false)
}

func (c *MigrateApplyCmd) Run(ctx context.Context, rt *app.Runtime) error {
	return migrate(ctx, rt, "apply", c.MigrationFile, true)
}

func migrate(ctx context.Context, rt *app.Runtime, action string, migrationFile string, includeApply bool) error {
	s := newService(rt)
	if err := s.beforeWorkspaceCommand(ctx); err != nil {
		return err
	}
	env := map[string]string{
		"TFMIGRATE_LOG":      "DEBUG",
		"TF_CLI_ARGS_plan":   strings.Join(s.varFileArgs(), " "),
		"TF_CLI_ARGS_import": strings.Join(s.varFileArgs(), " "),
	}
	if includeApply {
		env["TF_CLI_ARGS_apply"] = strings.Join(s.varFileArgs(), " ")
	}
	if err := s.runner.Run(ctx, s.workspaceSelect()); err != nil {
		return err
	}
	return s.runner.Run(ctx, runner.Command{Argv: []string{"tfmigrate", action, migrationFile}, Env: env})
}

func newService(rt *app.Runtime) *service {
	stdin := rt.Stdin
	if stdin == nil {
		stdin = os.Stdin
	}
	stdout := rt.Stdout
	if stdout == nil {
		stdout = os.Stdout
	}
	stderr := rt.Stderr
	if stderr == nil {
		stderr = os.Stderr
	}
	return &service{
		env:        rt.Target,
		verbosity:  rt.Verbosity,
		skipPrompt: rt.SkipPrompt,
		runner:     rt.Runner,
		stdin:      stdin,
		stdout:     stdout,
		stderr:     stderr,
		clock:      rt.Clock,
	}
}

func (s *service) beforeWorkspaceCommand(ctx context.Context) error {
	if err := s.tfenvInit(ctx); err != nil {
		return err
	}
	if err := s.checkEnv(); err != nil {
		return err
	}
	return s.promptEnvSwitch(ctx)
}

func (s *service) tfenvInit(ctx context.Context) error {
	if _, err := os.Stat(".terraform-version"); errors.Is(err, os.ErrNotExist) {
		fmt.Fprintln(s.stderr)
		fmt.Fprintln(s.stderr, "WARNING: .terraform-version not found -- using the default installed terraform")
		fmt.Fprintln(s.stderr)
		return nil
	} else if err != nil {
		return err
	}
	return s.sequence(ctx, tfenvInstallCmd())
}

func tfenvInstallCmd() runner.Command {
	return runner.Cmd("flock", tfenvInstallLockPath, "tfenv", "install")
}

func (s *service) checkEnv() error {
	if _, err := os.Stat(filepath.Join("vars", s.env)); errors.Is(err, os.ErrNotExist) {
		fmt.Fprintf(s.stderr, "\nNo environment in this project: %s\n\n.", s.env)
		return runner.Exit(1)
	} else if err != nil {
		return err
	}
	return nil
}

func (s *service) promptEnvSwitch(ctx context.Context) error {
	out, err := s.runner.Capture(ctx, runner.Cmd("terraform", "workspace", "show"))
	if err != nil {
		return err
	}
	currEnv := strings.TrimSpace(string(out))
	if currEnv == s.env {
		return nil
	}
	fmt.Fprintf(s.stderr, "\nswitching environment \x1b[31m%s\x1b[0m ~> \x1b[32m%s\x1b[0m\n\n", currEnv, s.env)
	if s.skipPrompt {
		return nil
	}
	reader := bufio.NewReader(s.stdin)
	for {
		fmt.Fprintf(s.stderr, "switch to %s? [y/N] ", s.env)
		resp, err := reader.ReadString('\n')
		if err != nil && !errors.Is(err, io.EOF) {
			return err
		}
		resp = strings.ToLower(strings.TrimSpace(resp))
		if resp == "" {
			resp = "n"
		}
		if resp == "n" || resp == "no" {
			return runner.Exit(1)
		}
		if resp == "y" || resp == "yes" {
			return nil
		}
		fmt.Fprintln(s.stderr, "what?")
		if errors.Is(err, io.EOF) {
			return runner.Exit(1)
		}
	}
}

func (s *service) confirmApply(planPath string) bool {
	fmt.Fprintln(s.stdout)
	fmt.Fprintln(s.stdout, "Would you like to apply this plan?")
	reader := bufio.NewReader(s.stdin)
	cont := ""
	for cont != "yes" {
		fmt.Fprintln(s.stdout, "You must answer 'yes' to continue")
		fmt.Fprint(s.stdout, "> ")
		resp, err := reader.ReadString('\n')
		if err != nil && !errors.Is(err, io.EOF) {
			fmt.Fprintln(s.stdout)
			fmt.Fprintln(s.stdout, "Aborted -- plan will not be applied")
			fmt.Fprintf(s.stdout, "path: %s\n", planPath)
			return false
		}
		if errors.Is(err, io.EOF) && resp == "" {
			fmt.Fprintln(s.stdout)
			cont = "no"
		} else {
			cont = strings.TrimSpace(resp)
		}
		if cont == "no" {
			fmt.Fprintln(s.stdout, "Aborted -- plan will not be applied")
			fmt.Fprintf(s.stdout, "path: %s\n", planPath)
			return false
		}
	}
	fmt.Fprintln(s.stdout)
	return true
}

func (s *service) varFileArgs() []string {
	var args []string
	for _, path := range s.locateVarFiles() {
		args = append(args, "-var-file="+path)
	}
	return args
}

func (s *service) locateVarFiles() []string {
	var paths []string
	for _, env := range []string{"common", s.env} {
		matches, _ := filepath.Glob(filepath.Join("vars", env, "*.tfvars"))
		sort.Strings(matches)
		for _, match := range matches {
			paths = append(paths, match)
		}
	}
	if s.verbosity >= 1 {
		for _, path := range paths {
			rel := strings.TrimPrefix(path, "vars"+string(filepath.Separator))
			fmt.Fprintf(s.stdout, "Variable file: %s\n", rel)
		}
	}
	return paths
}

func (s *service) workspaceSelect() runner.Command {
	return runner.Cmd("terraform", "workspace", "select", s.env)
}

func (s *service) sequence(ctx context.Context, cmds ...runner.Command) error {
	return runner.Sequence(ctx, s.runner, s.stderr, cmds...)
}

type realClock struct{}

func (realClock) Unix() int64 { return time.Now().Unix() }
